xarray/core/computation.py

```diff
@@ -1327,7 +1327,23 @@ def _cov_corr(da_a, da_b, dim=None, ddof=0, method=None):
     # 2. Ignore the nans
     valid_values = da_a.notnull() & da_b.notnull()
 
-    if not valid_values.all():
+    def _nan_check(d):
+        if d.all():
+            return True
+        else:
+            return False
+
+    if is_duck_dask_array(valid_values.data):
+        # assign to copy - else the check is not triggered
+        _are_there_nans = valid_values.copy(
+            data=valid_values.data.map_blocks(_nan_check, dtype=valid_values.dtype),
+            deep=False,
+        )
+
+    else:
+        _are_there_nans = _nan_check(valid_values.data)
+
+    if not _are_there_nans:
```

Comment:
Thanks @AndrewWilliams3142 .

Unfortunately I don't think you can avoid this for dask arrays. Fundamentally `bool(dask_array)` will call compute. So this line `if not _are_there_nans` will call compute.

I think the only solution here is to special case this optimization for non-dask DataArray but maybe someone else has an idea.

Reply:
OK I guess you could stick the 
```
valid_values = ...
if valid_values.any():
    ...
```

in a helper function and `map_blocks` that.

Reply:
Hmm, ok. Silly question but does it matter that `_are_there_nans` is a scalar reduction of `valid_values`? So `valid_values` is still left uncomputed. Would this help the bottleneck at all?

Reply:
No because the program needs to know `valid_blocks.any()` to decide which branch of the if-else statement to execute. So it gets computed.